psi4/driver/json_wrapper.py

```diff
@@ -52,11 +55,126 @@
+        "scf_dispersion_correction_energy": {"variables": "DISPERSION CORRECTION ENERGY", "skip_zero": True},
+
+        # SCF Properties (experimental)
+        # "scf_quadrupole_moment": {"variables": ["SCF QUADRUPOLE " + x for x in ["XX", "XY", "XZ", "YY", "YZ", "ZZ"]], "skip_null": True},
+        # "scf_mulliken_charges": {"variables": "MULLIKEN_CHARGES", "skip_null": True},
+        # "scf_lowdin_charges": {"variables": "LOWDIN_CHARGES", "skip_null": True},
+        # "scf_wiberg_lowdin_indices": {"variables": "WIBERG_LOWDIN_INDICES", "skip_null": True},
+        # "scf_mayer_indices": {"variables": "MAYER_INDICES", "skip_null": True},
+    },
+
+    # MP2 variables
+    "mp2": {
+        "mp2_same_spin_correlation_energy": {"variables": "MP2 SAME-SPIN CORRELATION ENERGY"},
+        "mp2_opposite_spin_correlation_energy": {"variables": "MP2 OPPOSITE-SPIN CORRELATION ENERGY"},
+        "mp2_singles_energy": {"variables": "NYI", "default": 0.0},
+        "mp2_doubles_energy": {"variables": "MP2 CORRELATION ENERGY"},
+        "mp2_total_correlation_energy": {"variables": "MP2 CORRELATION ENERGY"},
+        "mp2_total_energy": {"variables": "MP2 TOTAL ENERGY"},
+    },
+
+#    "": {"variables": },
+
+} # yapf: disable
+
+def _json_translation(value):
+    """
+    Translates from Psi4 to JSON data types
+    """
+
+    if isinstance(value, (psi4.core.Matrix, psi4.core.Vector)):
+        value = value.np.ravel().tolist()
+    elif isinstance(value, (psi4.core.Dimension)):
+        value = value.to_tuple()
+    elif isinstance(value, np.ndarray):
+        value = value.ravel().tolist()
+
+    return value
+
+def _convert_variables(data, include=None):
+    """
+    Converts dictionaries of variables based on translation metadata
+    """
+
+    # Build the correct translation units
+    if include is None:
+        needed_vars = {}
+        for v in _qcschema_translation.values():
+            needed_vars.update(v)
+    else:
+        needed_vars = _qcschema_translation[include]
+
+    ret = {}
+    for key, var in needed_vars.items():
+
+        # Get the actual variables
+        if isinstance(var["variables"], str):
+            value = data.get(var["variables"], None)
+        elif isinstance(var["variables"], (list, tuple)):
+            value = [data.get(x, None) for x in var["variables"]]
+            if not any(value):
+                value = None
+        else:
+            raise TypeError("variables type not understood.")
+
+        # Handle skips
+        if var.get("skip_zero", False) and (value == 0):
+            continue
+
+        if (var.get("skip_zero") or var.get("skip_null", False)) and (value is None):
+            continue
+
+        # Add defaults
+        if (value is None) and ("default" in var):
+            value = var["default"]
+
+        # Cast if called
+        if "cast" in var:
+            value = var["cast"](value)
+
+        ret[key] = _json_translation(value)
+
+    return ret
```

Comment:
nice. will have to eventually hook it into the [validator](https://github.com/qcdb/qcdb/blob/master/qcdb/qcvars/psivardefs.py) and glossary that ensures psivars are consistent and meet normal QC definitions (ref + corl = total; same-spin + opposite-spin + singles = total).

Reply:
That is mostly my thought. We slowly push all of this into metadata so that we can write other manipulators and validators with ease.